File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> ans;
        long product = 1;
        int index;
        int isZero = 0;
        for(int i=0; i<nums.size();i++){
            ans.push_back(0);
            if(nums[i] == 0){
                index = i;
                isZero++;
                continue;
            }
            // else if(product == 0) product=1;
            product *= nums[i];
        }
        if(isZero>1) return ans;
        if(isZero==1){
            ans[index] = product;
            return ans;
        }
        for(int i=0; i<nums.size();i++){
            ans[i] = product/nums[i];
        }
        return ans;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int product = 1, n = nums.length, zeroCount = 0, zeroIndex = -1;
        int[] ans = new int[n];
        
        for(int i=0; i<n; i++){
            if(nums[i] == 0) {
                zeroCount ++;
                zeroIndex = i;
                continue;
            }
            product *= nums[i];
        }
        if(zeroCount>1) return ans;
        if(zeroCount == 1) {
            ans[zeroIndex] = product;
            return ans;
        }
        for(int i=0; i<n; i++){
            ans[i] = product/nums[i];
        }
        return ans;
    }
}